src/core/file.cc

```diff
@@ -1051,16 +1059,50 @@ static internal::alignments xfs_alignments(const dioattr& da, unsigned block_siz
 future<shared_ptr<file_impl>>
 make_file_impl(int fd, file_open_options options, int flags, struct stat st) noexcept {
     if (S_ISBLK(st.st_mode)) {
-        size_t block_size;
-        auto ret = ::ioctl(fd, BLKBSZGET, &block_size);
-        if (ret == -1) {
+        size_t logical_block_size;
+        if (auto ret = ::ioctl(fd, BLKBSZGET, &logical_block_size); ret == -1) {
             return make_exception_future<shared_ptr<file_impl>>(
                     std::system_error(errno, std::system_category(), "ioctl(BLKBSZGET) failed"));
         }
+        size_t physical_block_size;
+        if (auto ret = ::ioctl(fd, BLKPBSZGET, &physical_block_size); ret == -1) {
+            return make_exception_future<shared_ptr<file_impl>>(
+                    std::system_error(errno, std::system_category(), "ioctl(BLKPBSZGET) failed"));
+        }
+
+        // Query DIO memory alignment using statx (kernel 6.1+)
+        // This gives us the actual DMA buffer alignment requirement (typically 4 bytes for NVMe)
+        size_t memory_dma_alignment = physical_block_size; // fallback to physical_block_size
+        struct statx stx;
+        if (syscall(__NR_statx, fd, "", AT_EMPTY_PATH, STATX_DIOALIGN, &stx) == 0) {
+            if (stx.stx_mask & STATX_DIOALIGN) {
+#ifndef SEASTAR_STATX_NEEDS_DIO_FIELDS
+                // Use kernel-reported memory alignment (stx_dio_mem_align)
+                // Field is available in Linux 6.1+ headers
+                memory_dma_alignment = stx.stx_dio_mem_align;
+#else
+                // On older headers, the field doesn't exist in the struct
+                // The kernel may support it, but we can't access it safely
+                // Fall back to physical_block_size
```

Comment:
If SEASTAR_STATX_NEEDS_DIO_FIELDS is not defined, calling statx makes no sense at all

Reply:
should have been fixed.